File: bindings/R/FortMyrmidon/src/shapes.cpp
#include "shapes.h"

#include "eigen.h"

#include "Rcpp.h"

void fmCircle_show(const fort::myrmidon::Circle * c) {
	Rcpp::Rcout << "fmCircle ( center = "
	            << c->Center().transpose()
	            << ", radius = "
	            << c->Radius()
	            << " )\n";
}

fort::myrmidon::Circle fmCircle_create(const Eigen::Vector2d & center, double radius) {
	return fort::myrmidon::Circle(center,radius);
}

void fmCapsule_show(const fort::myrmidon::Capsule * c) {
	Rcpp::Rcout << "fmCapsule ( center1 = "
	            << c->C1().transpose()
	            << ", center2 = "
	            << c->C2().transpose()
	            << ", radius1 = "
	            << c->R1()
	            << ", radius2 = "
	            << c->R2()
	            << " )\n";
}

fort::myrmidon::Capsule fmCapsule_create(const Eigen::Vector2d & c1,
                                         const Eigen::Vector2d & c2,
                                         double r1,
                                         double r2) {
	return fort::myrmidon::Capsule(c1,c2,r1,r2);
}

void fmPolygon_show(const fort::myrmidon::Polygon * p) {
	Rcpp::Rcout << "fmPolygon (";
	for ( size_t i = 0 ; i < p->Size(); ++i ) {
		Rcpp::Rcout <<" (" << p->Vertex(i).transpose() << ")";
	}
	Rcpp::Rcout << " )\n";
}


fort::myrmidon::Polygon fmPolygon_create(const fort::myrmidon::Vector2dList & vertices) {
	return fort::myrmidon::Polygon(vertices);
}


RCPP_MODULE(shapes) {
	Rcpp::class_<fort::myrmidon::Shape>("fmShape");

	Rcpp::class_<fort::myrmidon::Circle>("fmCircle")
		.derives<fort::myrmidon::Shape>("fmShape")
		.constructor<Eigen::Vector2d,double>()
		.const_method("show",&fmCircle_show)
		.const_method("center",&fort::myrmidon::Circle::Center)
		.const_method("radius",&fort::myrmidon::Circle::Radius)
		.method("setCenter",&fort::myrmidon::Circle::SetCenter)
		.method("setRadius",&fort::myrmidon::Circle::SetRadius)
		;

	Rcpp::class_<fort::myrmidon::Capsule>("fmCapsule")
		.derives<fort::myrmidon::Shape>("fmShape")
		.constructor<Eigen::Vector2d,Eigen::Vector2d,double,double>()
		.const_method("show",&fmCapsule_show)
		.const_method("c1",&fort::myrmidon::Capsule::C1)
		.const_method("r1",&fort::myrmidon::Capsule::R1)
		.const_method("c2",&fort::myrmidon::Capsule::C2)
		.const_method("r2",&fort::myrmidon::Capsule::R2)
		.method("setC1",&fort::myrmidon::Capsule::SetC1)
		.method("setR1",&fort::myrmidon::Capsule::SetR1)
		.method("setC2",&fort::myrmidon::Capsule::SetC2)
		.method("setR2",&fort::myrmidon::Capsule::SetR2)
		;

	Rcpp::class_<fort::myrmidon::Polygon>("fmPolygon")
		.derives<fort::myrmidon::Shape>("fmShape")
		.constructor<fort::myrmidon::Vector2dList>()
		.const_method("show",&fmPolygon_show)
		.const_method("size",&fort::myrmidon::Polygon::Size)
		.const_method("vertex",&fort::myrmidon::Polygon::Vertex)
		.method("setVertex",&fort::myrmidon::Polygon::SetVertex)
		;


	Rcpp::function("fmCircleCreate", &fmCircle_create);
	Rcpp::function("fmCapsuleCreate", &fmCapsule_create);
	Rcpp::function("fmPolygonCreate", &fmPolygon_create);

}

namespace Rcpp {
template <> fort::myrmidon::Shape::ConstList as(SEXP exp) {
	using namespace fort::myrmidon;
	Rcpp::List geometries(exp);
	fort::myrmidon::Shape::ConstList newGeometry;
	for ( const auto & elem : geometries ) {
		try {
			auto c = Rcpp::as<Circle>(elem);
			newGeometry.push_back(std::make_shared<Circle>(c.ToPrivate()));
			continue;
		} catch (const std::exception & e) {

		}

		try {
			auto c = Rcpp::as<Capsule>(elem);
			newGeometry.push_back(std::make_shared<Capsule>(c.ToPrivate()));
			continue;
		} catch (const std::exception & e) {

		}

		try {
			auto p = Rcpp::as<Polygon>(elem);
			newGeometry.push_back(std::make_shared<Polygon>(p.ToPrivate()));
			continue;
		} catch (const std::exception & e) {

		}

		throw std::runtime_error("fort::myrmidon::Shape::ConstList must be list of fmCircle, fmCapsule and fmPolygon");
	}
	return newGeometry;
}

template <> SEXP wrap(const fort::myrmidon::Shape::ConstList & list) {
	Rcpp::List res;
	for ( const auto & shape : list ) {
		if ( auto c = std::dynamic_pointer_cast<const fort::myrmidon::Circle>(shape) ) {
			res.push_back(*c);
		}

		if ( auto c = std::dynamic_pointer_cast<const fort::myrmidon::Capsule>(shape) ) {
			res.push_back(*c);
		}

		if ( auto p = std::dynamic_pointer_cast<const fort::myrmidon::Polygon>(shape) ) {
			res.push_back(*p);
		}
	}
	return res;
}

}

File: bindings/R/FortMyrmidon/R/Shapes.R
#' @name fmShape
#' @title a Refrence class for any kind of shape
#' @description just defined for C++ jibber-jabber
NULL

#' @name fmCircle
#' @title a Reference Class that wraps a C++ \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Circle/fort::myrmidon::Circle}{fort::myrmidon::Circle}
#'
#' @field center
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Circle/fort::myrmidon::Circle::Center}{fort::myrmidon::Circle::Center}
#' @field radius
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Circle/fort::myrmidon::Circle::Radius}{fort::myrmidon::Circle::Radius}
#' @field setCenter
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Circle/fort::myrmidon::Circle::SetCenter}{fort::myrmidon::Circle::SetRadius}
#' @field setRadius
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Circle/fort::myrmidon::Circle::SetRadius}{fort::myrmidon::Circle::SetRadius}
#'
#' @examples
#' # creates a fmCircle of center 1,1 amnd radius 2
#' c <- fmCircleCreate(c(1,1),2)
#' # moves it to (0,0)
#' c$setCenter(c(0,0))
#'
NULL

#' @name fmCapsule
#' @title a Reference Class that wraps a C++ \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule}{fort::myrmidon::Capsule}
#'
#' @field c1
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::C1}{fort::myrmidon::Capsule::C1}
#' @field c2
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::C2}{fort::myrmidon::Capsule::C2}
#' @field r1
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::R1}{fort::myrmidon::Capsule::R1}
#' @field r2
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::R2}{fort::myrmidon::Capsule::R2}
#' @field setC1
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::SetC1}{fort::myrmidon::Capsule::SetC1}
#' @field setC2
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::SetC2}{fort::myrmidon::Capsule::SetC2}
#' @field setR1
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::SetR1}{fort::myrmidon::Capsule::SetR1}
#' @field setR2
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Capsule/fort::myrmidon::Capsule::SetR2}{fort::myrmidon::Capsule::SetR2}
#' @examples
#' # creates a fmCapsule of center (1,1) and (2,2) radius 0.5 and 0.4
#' c <- fmCapsuleCreate(c(1,1),c(2,2),0.5,0.4)
#' # moves second center to (0,0)
#' c$setC2(c(0,0))
NULL

#' @name fmPolygon
#' @title a Reference Class that wraps a C++ \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Polygon/fort::myrmidon::Capsule}{fort::myrmidon::Polygon}
#'
#' @field size
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Polygon/fort::myrmidon::Polygon::Size}{fort::myrmidon::Polygon::Size}
#' @field vertex
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Polygon/fort::myrmidon::Polygon::Vertex}{fort::myrmidon::Polygon::Vertex}
#' @field setVertex
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Polygon/fort::myrmidon::Polygon::SetVertex}{fort::myrmidon::Polygon::SetVertex}
#' @examples
#' #creates a nice (0,0) centered square of size 2
#' p <- fmPolygonCreate(list(c(-1,-1),c(1,-1),c(1,1),c(-1,1)))
#' # transforms it into the extinction rebellion symbol by swapping the
#' # two lastelements, as order matters here.
#' # ATTENTION  list index starts at 0, like in C
#' p$setVertex(2,c(-1,1))
#' p$setVertex(3,c(1,1))
NULL
